File: 376.摆动序列.cpp
/*
 * @lc app=leetcode.cn id=376 lang=cpp
 *
 * [376] 摆动序列
 */
#include <vector>
using namespace std;
// @lc code=start
class Solution
{
public:
    int wiggleMaxLength(vector<int> &nums)
    {
        if (nums.size() < 2)
            return nums.size();

        int up = 1, down = 1;
        // up,down表示当前元素作为up，down时的最长摆动序列

        for (int i = 1; i < nums.size(); i++)
        {
            if (nums[i] > nums[i - 1])
                up = down + 1;
            else if (nums[i] < nums[i - 1])
                down = up + 1;
            // 连续的上升或者下降是不会影响结果的，因为down或者up没变
        }
        return max(up, down);
    }
};
// @lc code=end

File: 376.摆动序列.py
#
# @lc app=leetcode.cn id=376 lang=python3
#
# [376] 摆动序列
#
from typing import List


# @lc code=start
class Solution:
    def wiggleMaxLength(self, nums: List[int]) -> int:
        # 连续上升：只保留最后那个最高点
        # 连续下降：只保留最后那个最低点
        # 方向反转：答案 +1

        prev_diff = 0
        res = 1
        for i in range(1, len(nums)):
            cur_diff = nums[i] - nums[i - 1]
            # 只有第一次遍历时prev_diff = 0, 后续均不为0
            if cur_diff * prev_diff <= 0 and cur_diff != 0:
                res += 1
                prev_diff = cur_diff

        return res


# @lc code=end
sol = Solution()
res = sol.wiggleMaxLength(nums = [1,2,3,4,5,6,7,8,9])
print(res)
